### ISSUE TITLE
[BUG] Snapshot restore after an upgrade the cluster gets stuck in an error state 

### ISSUE THREAD

**Issue by igomez06** [rancher/rancher#43222]
**Rancher Server Setup**
- Rancher version: v2.8.0-rc1
- Installation option (Docker install/Helm Chart): helm
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): v1.26.9+k3s1
- Proxy/Cert Details:

**Information about the Cluster**
- Kubernetes version:  v1.26.9+k3s1
- Cluster Type (Local/Downstream): Downstream
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider):
 AWS Node provider 2 etcd, 3 cp, 2 worker

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
  - If custom, define the set of permissions: Admin

**Describe the bug**
 When attempting to a restore a snapshot after the cluster had been upgraded to v1.27.6+k3s1 the cluster gets stuck in an Error state with the error message reading: 
`Unable to find and decode snapshot ClusterSpec for snapshot`

Some logs
```
Deleting etcd snapshot fleet-default/auto-aws-gnymn-on-demand-auto-aws-gnymn-pool0-99db806f-4q-0cc7c
2023-10-19T21:16:32.753634950Z 2023/10/19 21:16:32 [INFO] [plansecret] Deleting etcd snapshot fleet-default/auto-aws-gnymn-on-demand-auto-aws-gnymn-pool0-99db806f-8h-81772
```

```
 refreshing etcd create state
2023-10-19T21:16:29.880629557Z 2023/10/19 21:16:29 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: starting etcd snapshot on node auto-aws-gnymn-pool0-99db806f-4qgnp, starting etcd snapshot on node auto-aws-gnymn-pool0-99db806f-8hn4n
2023-10-19T21:16:30.295008355Z 2023/10/19 21:16:30 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: waiting for etcd snapshot on node auto-aws-gnymn-pool0-99db806f-4qgnp, waiting for etcd snapshot on node auto-aws-gnymn-pool0-99db806f-8hn4n
2023-10-19T21:16:30.454569426Z 2023/10/19 21:16:30 [INFO] [planner] rkecl
uster fleet-default/auto-aws-gnymn: waiting for etcd snapshot on node auto-aws-gnymn-pool0-99db806f-4qgnp, waiting for etcd snapshot on node auto-aws-gnymn-pool0-99db806f-8hn4n
2023-10-19T21:16:30.698137438Z 2023/10/19 21:16:30 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: waiting for etcd snapshot on node auto-aws-gnymn-pool0-99db806f-4qgnp, waiting for etcd snapshot on node auto-aws-gnymn-pool0-99db806f-8hn4n
2023-10-19T21:16:32.715818889Z 2023/10/19 21:16:32 [INFO] [plansecret] Deleting etcd snapshot fleet-default/auto-aws-gnymn-on-demand-auto-aws-gnymn-pool0-99db806f-4q-0cc7c
2023-10-19T21:16:32.753634950Z 2023/10/19 21:16:32 [INFO] [plansecret] Deleting etcd snapshot fleet-default/auto-aws-gnymn-on-demand-auto-aws-gnymn-pool0-99db806f-8h-81772
2023-10-19T21:16:32.900905042Z 2023/10/19 21:16:32 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: waiting for etcd snapshot on node auto-aws-gnymn-pool0-99db806f-8hn4n
2023-10-19T21:16:33.054112741Z 2023/10/19 21:16:33 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: refreshing etcd create state
2023-10-19T21:16:33.225957921Z 2023/10/19 21:16:33 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: configuring bootstrap node(s) auto-aws-gnymn-pool0-5c789784c9x6sghs-nhktd: waiting for plan to be applied
2023-10-19T21:16:33.324774934Z 2023/10/19 21:16:33 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: configuring bootstrap node(s) auto-aws-gnymn-pool0-5c789784c9x6sghs-nhktd: waiting for plan to be applied
2023-10-19T21:16:33.392109268Z 2023/10/19 21:16:33 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: configuring bootstrap node(s) auto-aws-gnymn-pool0-5c789784c9x6sghs-nhktd: waiting for plan to be applied
2023-10-19T21:16:33.558161004Z 2023/10/19 21:16:33 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: configuring bootstrap node(s) auto-aws-gnymn-pool0-5c789784c9x6sghs-nhktd: waiting for plan to be applied
2023-10-19T21:16:36.351299690Z 2023/10/19 21:16:36 [INFO] [planner] rkecluster fleet-default/auto-aw
s-gnymn: configuring bootstrap node(s) auto-aws-gnymn-pool0-5c789784c9x6sghs-nhktd: waiting for plan to be applied
2023-10-19T21:16:36.429503619Z 2023/10/19 21:16:36 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: auto-aws-gnymn-pool0-5c789784c9x6sghs-nhktd
2023-10-19T21:16:36.573560995Z 2023/10/19 21:16:36 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: starting etcd snapshot creation management plane restart
2023-10-19T21:16:36.675035847Z 2023/10/19 21:16:36 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: waiting for etcd snapshot creation management plane restart
2023-10-19T21:16:36.783696824Z 2023/10/19 21:16:36 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: waiting for etcd snapshot creation management plane restart
2023-10-19T21:16:36.852913738Z 2023/10/19 21:16:36 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: waiting for etcd snapshot creation management plane restart
2023-10-19T21:16:39.312946843Z 2023/10/19 21:16:39 [INFO] [planner] rkecluster fleet-default/auto-aws-gnymn: refreshing etcd create state
2023-10-19T21:16:39.423954545Z 2023/10/19 21:16:39 [ERROR] [planner] rkecluster fleet-default/auto-aws-gnymn: error parsing snapshot cluster spec for snapshot fleet-default/auto-aws-gnymn-on-demand-auto-aws-gnymn-pool0-99db806f-4q-0cc7c during etcd restoration: unable to find and decode snapshot ClusterSpec for snapshot
2023-10-19T21:16:39.559512953Z 2023/10/19 21:16:39 [ERROR] [planner] rkecluster fleet-default/auto-aws-gnymn: error parsing snapshot cluster spec for snapshot fleet-default/auto-aws-gnymn-on-demand-auto-aws-gnymn-pool0-99db806f-4q-0cc7c during etcd restoration: unable to find and decode snapshot ClusterSpec for snapshot
2023-10-19T21:16:39.687934507Z 2023/10/19 21:16:39 [ERROR] [planner] rkecluster fleet-default/auto-aws-gnymn: error parsing snapshot cluster spec for snapshot fleet-default/auto-aws-gnymn-on-demand-auto-aws-gnymn-pool0-99db806f-4q-0cc7c during etcd restoration: unable to find and decode snapshot ClusterSpec for snapshot
```

**To Reproduce**

A provision a 2 etcd, 3 cp, 2 worker k3s cluster version v1.26.9+k3s1, run go automation test `TestSnapshotRestore/TestSnapshotRestoreWithK8sUpgrade`

**Result**
The test will timeout, since it never goes back to an active state.
**Expected Result**

For the cluster to restored to its previous state and version. 
**Screenshots**

**Additional context**

**Comment by brandond** [comment]
I suspect that this is may be related to the issue of the snapshot metadata configmap exceeding the max size, and the snapshot metadata not being available? This is a long-standing issue that will be resolved in the October releases of K3s and RKE2. I am not sure why we're still doing Rancher validations on the August or September releases (v1.26.8/v1.26.9) as we have already agreed that these will be skipped entirely in KDM.

cc @rancher-max @snasovich

**Comment by igomez06** [comment]
I was testing v1.26.9, because that's what the kdm dev-v2.8 branch had as the latest 1.26 rke2 version. Which is what the latest v2.8 Rancher rc had in its default 1.26 k3s version. I wasn't aware of what was agreed upon.

**Comment by ElanHasson** [comment]
I hit this also. 

` rke2 etcd-snapshot prune --name $prefixofbackuptodelete` helped me shrink the configmap it so it can be used to restore.

**Comment by brandond** [comment]
This will be fixed in the October releases of rke2.

**Comment by ElanHasson** [comment]
As always, thanks Brandon!

**Comment by rancher-max** [comment]
This has been validated using rancher v2.8.0-rc3 with the latest k3s versions available: v1.26.10+k3s2 --> upgrade to v1.27.7+k3s2 --> restore snapshot back to v1.26.10+k3s2. We have also run it multiple times on both k3s and rke2 in nightly runs and seen no failures.

Below are some screenshots captured during a test and a report:

- Upgrading to v1.27.7+k3s2:
<img width="1317" alt="image" src="https://github.com/rancher/rancher/assets/55997940/fab76a0d-73b4-4008-8016-6b160449ac1e">
- Restore after upgrade:
<img width="1336" alt="image" src="https://github.com/rancher/rancher/assets/55997940/200b7fb7-8298-4f85-916e-5df5a2f4d9d4">

- Provisioning log:
```
[INFO ] waiting for at least one control plane, etcd, and worker node to be registered
[INFO ] waiting for viable init node
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-zhtdx: waiting for agent to check in and apply initial plan
[INFO ] configuring etcd node(s) maxk126-etcd-54c67756ddxklc8w-wtjrp: waiting for agent to check in and apply initial plan
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-dllwj,maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] waiting for control plane to be available
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-dhzdh,maxk126-worker-85db9794d5xq98c6-vxwd9
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-dhzdh: waiting for agent to check in and apply initial plan
[INFO ] provisioning done
[INFO ] waiting for etcd snapshot on node maxk126-etcd-80399a4d-hgdjw, waiting for etcd snapshot on node maxk126-etcd-80399a4d-8gz5s
[INFO ] waiting for etcd snapshot on node maxk126-etcd-80399a4d-hgdjw
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-zhtdx: waiting for plan to be applied
[INFO ] maxk126-etcd-54c67756ddxklc8w-zhtdx
[INFO ] waiting for etcd snapshot creation management plane restart
[INFO ] provisioning done
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-zhtdx: waiting for plan to be applied
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-zhtdx: waiting for kubelet to update
[INFO ] maxk126-etcd-54c67756ddxklc8w-zhtdx
[INFO ] configuring etcd node(s) maxk126-etcd-54c67756ddxklc8w-wtjrp: waiting for plan to be applied
[INFO ] configuring etcd node(s) maxk126-etcd-54c67756ddxklc8w-wtjrp: waiting for kubelet to update
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-dllwj,maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-mwhs6: waiting for plan to be applied
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-mwhs6: waiting for kubelet to update
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-dhzdh,maxk126-worker-85db9794d5xq98c6-vxwd9
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-vxwd9: waiting for kubelet to update
[INFO ] provisioning done
[INFO ] waiting to stop k3s services on node [maxk126-cp-b366e045-wgghf]
[INFO ] waiting to stop k3s services on node [maxk126-worker-98747167-zcz59]
[INFO ] waiting for etcd restore
[INFO ] refreshing etcd restore state
[INFO ] configuring etcd node(s) maxk126-etcd-54c67756ddxklc8w-zhtdx: waiting for plan to be applied
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-wtjrp: waiting for kubelet to update
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-dllwj,maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-mwhs6: Node condition MemoryPressure is Unknown. Node condition DiskPressure is Unknown. Node condition PIDPressure is Unknown. Node condition Ready is Unknown., waiting for plan to be applied
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-mwhs6: Node condition Ready is False., waiting for plan to be applied
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-mwhs6: waiting for probes: kube-controller-manager, kube-scheduler
[INFO ] maxk126-cp-5d7f96bd88xjgkl6-dllwj,maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-dhzdh,maxk126-worker-85db9794d5xq98c6-vxwd9
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-vxwd9: Node condition MemoryPressure is Unknown. Node condition DiskPressure is Unknown. Node condition PIDPressure is Unknown. Node condition Ready is Unknown., waiting for plan to be applied
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-vxwd9: waiting for plan to be applied
[INFO ] waiting for etcd restore
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-wtjrp: waiting for plan to be applied
[INFO ] provisioning done
[INFO ] waiting for etcd snapshot on node maxk126-etcd-80399a4d-hgdjw, waiting for etcd snapshot on node maxk126-etcd-80399a4d-8gz5s
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-wtjrp: waiting for plan to be applied
[INFO ] waiting for etcd snapshot creation management plane restart
[INFO ] provisioning done
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-wtjrp: waiting for plan to be applied
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-wtjrp: waiting for kubelet to update
[INFO ] configuring etcd node(s) maxk126-etcd-54c67756ddxklc8w-zhtdx: waiting for plan to be applied
[INFO ] configuring etcd node(s) maxk126-etcd-54c67756ddxklc8w-zhtdx: waiting for kubelet to update
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-dllwj,maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-dllwj,maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-dhzdh,maxk126-worker-85db9794d5xq98c6-vxwd9
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-vxwd9: waiting for plan to be applied
[INFO ] provisioning done
[INFO ] waiting to stop k3s services on node [maxk126-cp-b366e045-wgghf]
[INFO ] waiting for etcd restore
[INFO ] configuring etcd node(s) maxk126-etcd-54c67756ddxklc8w-zhtdx: waiting for plan to be applied
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-wtjrp: waiting for kubelet to update
[INFO ] maxk126-etcd-54c67756ddxklc8w-zhtdx
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-dllwj,maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-h4525,maxk126-cp-5d7f96bd88xjgkl6-mwhs6
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-mwhs6: Node condition MemoryPressure is Unknown. Node condition DiskPressure is Unknown. Node condition PIDPressure is Unknown. Node condition Ready is Unknown., waiting for plan to be applied
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-mwhs6: Node condition Ready is False., waiting for plan to be applied
[INFO ] configuring control plane node(s) maxk126-cp-5d7f96bd88xjgkl6-mwhs6: waiting for plan to be applied
[INFO ] marking control plane as initialized and ready
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-dhzdh,maxk126-worker-85db9794d5xq98c6-vxwd9
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-vxwd9: Node condition MemoryPressure is Unknown. Node condition DiskPressure is Unknown. Node condition PIDPressure is Unknown. Node condition Ready is Unknown., waiting for plan to be applied
[INFO ] configuring worker node(s) maxk126-worker-85db9794d5xq98c6-vxwd9: Node condition Ready is False., waiting for plan to be applied
[INFO ] waiting for etcd restore
[INFO ] configuring bootstrap node(s) maxk126-etcd-54c67756ddxklc8w-wtjrp: waiting for plan to be applied
[INFO ] provisioning done
```

- Automated test result:
<img width="1260" alt="image" src="https://github.com/rancher/rancher/assets/55997940/72465542-79d8-4bfd-a287-e12ff6761a63">
